hibernate-core/src/main/java/org/hibernate/query/sqm/consume/spi/BaseSqmToSqlAstConverter.java

```diff
@@ -1013,25 +1112,62 @@ public void accept(SqlExpressableType type) {
 
 	@Override
 	public Object visitJpaParameterWrapper(JpaParameterSqmWrapper expression) {
+		log.tracef( "#visitJpaParameterWrapper(%s)", expression );
+
+		final AllowableParameterType<?> resolvedType = resolveParameterType(
+				expression,
+				() -> {
+					final ParameterExpression<?> criteriaParamExpr = expression.getJpaParameterExpression();
+
+					final AllowableParameterType explicitType = criteriaParamExpr.getExplicitType();
+					if ( explicitType != null ) {
+						return explicitType;
+					}
+
+					final AllowableParameterType<?> hibernateType = criteriaParamExpr.getHibernateType();
+					if ( hibernateType != null ) {
+						return hibernateType;
+					}
+
+					final Class<?> javaType = criteriaParamExpr.getJavaType();
+					if ( javaType != null ) {
+						final TypeConfiguration typeConfiguration = getSessionFactory().getTypeConfiguration();
+						try {
+							final BasicValuedExpressableType<?> type = typeConfiguration.standardExpressableTypeForJavaType( javaType );
+							if ( type != null ) {
+								return type;
+							}
+						}
+						catch (Exception e) {
+							log.debugf( "Unable to resolve Java type [%s] to AllowableParameterType : %s", javaType.getName(), e );
+						}
+					}
+
+					throw new IllegalArgumentException( "Could not resolve AllowableParameterType for JpaParameterSqmWrapper : " + expression );
+				}
+		);
+
+		registerParameterResolvedType( expression, resolvedType );
+
+		final List<JdbcParameter> existing = this.jdbcParamsBySqmParam.get( expression );
 		final List<JdbcParameter> jdbcParameterList;
 
-		List<JdbcParameter> existing = this.jdbcParamsBySqmParam.get( expression );
 		if ( existing != null ) {
-			final int number = expression.getExpressableType().getNumberOfJdbcParametersNeeded();
+			final int number = resolvedType.getNumberOfJdbcParametersNeeded();
 			assert existing.size() == number;
 			jdbcParameterList = existing;
 		}
 		else {
 			jdbcParameterList = new ArrayList<>();
 
 			//noinspection Convert2Lambda
-			expression.getExpressableType().visitJdbcTypes(
+			resolvedType.visitJdbcTypes(
 					new Consumer<SqlExpressableType>() {
 						@Override
 						public void accept(SqlExpressableType type) {
 							jdbcParameterList.add(
 									new StandardJdbcParameterImpl(
-											jdbcParameters.getJdbcParameters().size(),
+											jdbcParameters.getJdbcParameters().size() + jdbcParameterList.size() + 1,
```

Comment:
Actually that integer is never used "for real".  It is literally only ever used to provide uniqueness to compare 2 different `StandardJdbcParameterImpl` impls internally (they are comparable).

So I could literally set it to `jdbcParameterList.size() + 9999999999` and have the same affect ;)

It is another thing I had planned to come back and address but forgot.  I'll address this too

Reply:
To clarify...this is not the "position" used in the actual JDBC param binding.

Reply:
But for what it is worth... I added `+1` because at that point in time the parameter has not yet been added to `jdbcParameterList`.  The idea was to replicate as closely as possible the parameter index ultimately used in JDBC param binding later - which of course is 1 based..